File: services/app/utils/publish_to_topic.py
from confluent_kafka import Producer
import json
from pathlib import Path

# Get the path to the root directory
root_dir = Path(__file__).resolve().parent.parent

properties_file = root_dir / "client.properties"

def read_config():
  # reads the client configuration from client.properties
  # and returns it as a key-value map
  config = {}
  with open(properties_file) as fh:
    for line in fh:
      line = line.strip()
      if len(line) != 0 and line[0] != "#":
        parameter, value = line.strip().split('=', 1)
        config[parameter] = value.strip()
  return config

def produce(topic, data):
  # creates a new producer instance
  producer = Producer(read_config())

  # produces a sample message
  producer.produce(topic, value=json.dumps(data))

  # send any outstanding or buffered messages to the Kafka broker
  producer.flush()

File: web-application/util/publish-to-topic.js
const { Kafka } = require("@confluentinc/kafka-javascript").KafkaJS;
const fs = require('fs');

// Configuration for for Kafka
const config = readConfig(process.cwd() + "/client.properties");

// Reads the Kafka config file
function readConfig(fileName) {
  const data = fs.readFileSync(fileName, "utf8").toString().split("\n");

  return data.reduce((config, line) => {
    const [key, value] = line.split("=");
    if (key && value) {
      config[key] = value;
    }
    return config;
  }, {});
}

module.exports = {
  // Saves the data object to a Kafka topic
  publishToTopic: async function(topic, data) {
    console.log('writeToTopic');

    // Create a new producer instance
    const producer = new Kafka().producer(config);

    // Connect the producer to the broker
    await producer.connect();

    // Convert the text chunks array into an array that can be sent to Kafka
    const messages = data.map(item => ({ value: JSON.stringify(item) }));

    // Send a single message
    const produceRecord = await producer.send({
      topic,
      messages: messages
    });
    // console.log(
    //   `\n\n Produced message to topic ${topic}: , ${JSON.stringify(
    //     produceRecord,
    //     null,
    //     2
    //   )} \n\n`
    // );

    // Disconnect the producer
    await producer.disconnect();
  }
}